mmseg/models/losses/phi_loss.py

```diff
@@ -0,0 +1,90 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from ..builder import LOSSES
+from .utils import reduce_loss
+from .cross_entropy_loss import _expand_onehot_labels
+
+
+def phi_loss(inputs, targets, gamma, smooth, weight=None, reduction='mean'):
+    
+    inputs = inputs.sigmoid()
+    targets= targets.type_as(inputs)
+    
+    # flatten label and prediction tensors
+    inputs = inputs.reshape(inputs.size(0), -1)
+    targets = targets.reshape(targets.size(0), -1)
+    weight = weight.reshape(weight.size(0), -1)
+    
+    TP = (inputs * targets).sum()
+    FP = ((1 - targets) * inputs).sum()
+    FN = (targets * (1 - inputs)).sum()
+    TN= inputs[targets==0.].sum()
+    
+    phi = ((TP * TN - FP * FN) + smooth) / torch.sqrt((TP + FP) * (TP + FN) * (TN + FP) * (TN + FN) + smooth) 
+    loss= (1. - phi) ** gamma #focal_phi_loss
+    loss= reduce_loss(loss,reduction=reduction)
+    return loss
+
+
+@LOSSES.register_module()
+class PhiLoss(nn.Module):
+    """PhiLoss.
+
+    Args:
+        reduction (str, optional): . Defaults to 'mean'.
+            Options are "none", "mean" and "sum".
+        smooth (float, optional): Smoothing factor. Defaults to 1.0.
+        loss_weight (float, optional): Weight of the loss. Defaults to 1.0.
+    """
+
+    def __init__(self,
+    			 gamma= 1.0,
+                 reduction='mean',
+                 smooth=1,
+                 loss_weight=1.0,
+                 loss_name='phi_loss'):
+        super(PhiLoss, self).__init__()
+        self.gamma = gamma
+        self.reduction = reduction
+        self.loss_weight = loss_weight
+        self.smooth = smooth
+        self._loss_name = loss_name
+
+    def forward(self,
+                inputs,
+                targets,
+                weight=None,
+                reduction_override=None,
+                ignore_index=255):
+        """Forward function."""
```

Comment:
I suggest add the docstring for these arguments to explain their function

Reply:
Hi, the function is indeed identical to the original that you have at BCE:

https://github.com/open-mmlab/mmsegmentation/blob/ee25adc2c1d73e22355d168fa09ae7bdac899a24/mmseg/models/losses/cross_entropy_loss.py#L254

The function could be understood by those who want to do it from checking the ` _expand_onehot_labels` function in the original BCE:

https://github.com/open-mmlab/mmsegmentation/blob/ee25adc2c1d73e22355d168fa09ae7bdac899a24/mmseg/models/losses/cross_entropy_loss.py#L66

Honestly, I would love to provide a description for what's happening, but I'm not able to as I am not the one who developed these parts.

Reply:
ok I'll add the docstring for all `forward` function after this pr merged